mlflow/johnsnowlabs.py

```diff
@@ -0,0 +1,786 @@
+                                      _PYTHON_ENV_FILE_NAME,
+                                      _REQUIREMENTS_FILE_NAME,
+                                      _mlflow_conda_env, _process_conda_env,
+                                      _process_pip_requirements, _PythonEnv)
+from mlflow.utils.file_utils import (TempDir,
+                                     shutil_copytree_without_file_permissions,
+                                     write_to)
+from mlflow.utils.model_utils import (_add_code_from_conf_to_system_path,
+                                      _get_flavor_configuration_from_uri,
+                                      _validate_and_copy_code_paths)
+from mlflow.utils.uri import (append_to_uri_path, dbfs_hdfs_uri_to_fuse_path,
+                              generate_tmp_dfs_path,
+                              get_databricks_profile_uri_from_artifact_uri,
+                              is_local_uri, is_valid_dbfs_uri)
+
+FLAVOR_NAME = "johnsnowlabs"
+_JOHNSNOWLABS_ENV_VARS = ('SECRET', 'AWS_ACCESS_KEY_ID', 'AWS_SECRET_ACCESS_KEY', 'SPARK_NLP_LICENSE')
+_JOHNSNOWLABS_MODEL_PATH_SUB = "jsl-model"
+_MLFLOWDBFS_SCHEME = "mlflowdbfs"
+_SPARK_NLP_JAR_URI = f"https://s3.amazonaws.com/auxdata.johnsnowlabs.com/public/jars/spark-nlp-assembly-{settings.raw_version_nlp}.jar"
+_SPARK_NLP_JSL_JAR_URI = "https://pypi.johnsnowlabs.com/{secret}/spark-nlp-jsl-" + f"{settings.raw_version_medical}.jar"
+_SPARK_NLP_JSL_WHEEL_URI = "https://pypi.johnsnowlabs.com/{secret}/spark-nlp-jsl/spark_nlp_jsl-" + f"{settings.raw_version_medical}-py3-none-any.whl"
+_logger = logging.getLogger(__name__)
+
+
+def _validate_env_vars():
+    if not all(os.environ.get(var) for var in _JOHNSNOWLABS_ENV_VARS):
+        raise Exception(
+            f'Please set {",".join(var for var in _JOHNSNOWLABS_ENV_VARS if not os.environ.get(var))}'
+            f' environment variables to your secret key before saving or logging your model')
+
+
+def _download_url(url, save_path):
+    file_name = url.split("/")[-1]
+    save_path = Path(save_path) / file_name
+    # Download the file from `url` and save it locally under `file_name`:
+    with urlopen(url) as response, open(str(save_path), "wb") as out_file:
+        shutil.copyfileobj(response, out_file)
+
+
+def _download_sparknlp_jar(out_folder):
+    _download_url(
+        _SPARK_NLP_JAR_URI,
+        out_folder)
+
+
+def _download_sparknlp_jsl_jar(out_folder, secret):
+    _download_url(
+        _SPARK_NLP_JSL_JAR_URI.format(secret=secret),
+        out_folder)
+
+
+def get_default_pip_requirements():
+    """
+    :return: A list of default pip requirements for MLflow Models produced by this flavor.
+             Calls to :func:`save_model()` and :func:`log_model()` produce a pip environment
+             that, at minimum, contains these requirements.
+    """
+    _validate_env_vars()
+    import pyspark
+    return [
+        f'johnsnowlabs_for_databricks=={settings.raw_version_jsl_lib}',
```

Comment:
What is the difference between https://pypi.org/project/johnsnowlabs/ and https://pypi.org/project/johnsnowlabs-for-databricks?

Reply:
There is only 1 difference, in the `setup.py` 
https://pypistats.org/packages/johnsnowlabs depends on a **fixed pyspark version** (our recommended version)

https://pypistats.org/packages/johnsnowlabs-for-databricks does **not depend on pyspark** 
everything else is exactly the same codebase. 

----

Example why we need it here: 

Our current recommended pyspark version is 3.1.2
A user may want to save a model using pyspark 3.3.0.


Since we want to save the same pyspark version as was used when the model is saved, we write `pyspark==3.3.0` to requirements.txt. Now if we write `johnsnowlabs` to the requirements.txt we will have a dependency resolution conflict from pip when creating a venv/container for `mlflow serve`. 


The only way to circumvent it for now is using `johnsnowlabs_tmp` which removes the constraint/dependency on pyspark

Reply:
> Our current recommended pyspark version is 3.1.2

Does this mean `johnsnowlabs` may not work with pyspark > 3.1.2?